In [27]:
!pip install transformers[sentencepiece]

In [ ]:
from transformers import pipeline

Sentiment Analysis

In [ ]:
classifier = pipeline('sentiment-analysis')

No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision af0f99b (https://huggingface.co/distilbert/distilbert-base-uncased-finetuned-sst-2-english).
Using a pipeline without specifying a model name and revision in production is not recommended.
/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

The pipeline API chooses a fine-tuned, pre-trained model for the purpose of sentiment analysis in English. We downloaded this model and cached it after creating the classifier object. Therefore, after rerunning the above command, the cached model will be loaded instead of being downloaded.

We can verify which model the present classifier object is using in the following way:

In [ ]:
classifier.model.name_or_path

'distilbert/distilbert-base-uncased-finetuned-sst-2-english'

Here, the classifier chooses a type of BERT model

In [ ]:
classifier("I am  really excited about about today !! ")

[{'label': 'POSITIVE', 'score': 0.9997952580451965}]

We observe that the classifier has classified the input sentence as a POSITIVE sentence with almost 100% confidence

In [ ]:
classifier(["I am very excited for this new movie !!",
            "I am very happy",
            "I hate this weather !!",
            "I really hate that movie.."])

[{'label': 'POSITIVE', 'score': 0.9998107552528381},
 {'label': 'POSITIVE', 'score': 0.9998795986175537},
 {'label': 'NEGATIVE', 'score': 0.9992252588272095},
 {'label': 'NEGATIVE', 'score': 0.9994947910308838}]

With a few more examples, we have verified that the classifier has classified the sentiment of the sentences correctly with good confidence

**Zero-Shot Classification**

In Zero-Shot classification, the input texts are not labeled. Here, we need to define the labels as per our needs

In [ ]:
from transformers import pipeline
classifier = pipeline('zero-shot-classification')

No model was supplied, defaulted to facebook/bart-large-mnli and revision c626438 (https://huggingface.co/facebook/bart-large-mnli).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.15k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

In [ ]:
classifier("This is really a very good course about how to cook mutton",
           candidate_labels = ["Education", "Eat", "Cooking"])

{'sequence': 'This is really a very good course about how to cook mutton',
 'labels': ['Cooking', 'Eat', 'Education'],
 'scores': [0.7549448609352112, 0.2265607714653015, 0.01849440671503544]}

It is almost 75% confident that the sentence is about cooking

In [ ]:
classifier(
    ["This is a course on advanced deep learning",
     "This AI course can generate more than 10 million in revenue",],
    candidate_labels=["education", "ai", "business"],
)

[{'sequence': 'This is a course on advanced deep learning',
  'labels': ['ai', 'education', 'business'],
  'scores': [0.7214259505271912, 0.21009045839309692, 0.06848355382680893]},
 {'sequence': 'This AI course can generate more than 10 million in revenue',
  'labels': ['ai', 'business', 'education'],
  'scores': [0.6334292888641357, 0.3411827087402344, 0.02538798376917839]}]

It is almost 72% confident that the first sentence is about AI and almost 63% sure that the second sentence is about AI

**Text Generation**

The text generation is done using the initial prompt, and the model auto-completes the remaining text. However, text generation involves some randomness, and the results may not match exactly

In [ ]:
from transformers import pipeline

generator = pipeline('text-generation')

No model was supplied, defaulted to openai-community/gpt2 and revision 6c0e608 (https://huggingface.co/openai-community/gpt2).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

In [ ]:
results = generator("I am really happy because ",
          num_return_sequences=2,
          max_length=30)

Truncation was not explicitly activated but `max_length` is provided a specific value, please use `truncation=True` to explicitly truncate examples to max length. Defaulting to 'longest_first' truncation strategy. If you encode pairs of sequences (GLUE-style) with the tokenizer you can select this strategy more precisely by providing a specific strategy to `truncation`.
Setting `pad_token_id` to `eos_token_id`:50256 for open-end generation.


In [ ]:
for i in results:
    print(i['generated_text'])
    print('\n')

I am really happy because  I don't know if she really knows. However, it is a very small number of people that do not know


I am really happy because  I don't feel like going there and getting fired.
Now for my final thoughts on the story...
I




**Question Answering**

The question answering pipeline can answer questions by understanding the context of the given information

In [ ]:
from transformers import pipeline

question_answerer = pipeline("question-answering")

No model was supplied, defaulted to distilbert/distilbert-base-cased-distilled-squad and revision 626af31 (https://huggingface.co/distilbert/distilbert-base-cased-distilled-squad).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/473 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/261M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/29.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/213k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/436k [00:00<?, ?B/s]

In [ ]:
question_answerer(
    question = "What is the capital of India ?",
    context = """India, officially the Republic of India, is a country in South Asia.
    It is the second-most populous country, the seventh-largest country by land area, and the most populous democracy in the world.
    The capital is New Delhi"""
)

{'score': 0.9928799867630005, 'start': 220, 'end': 229, 'answer': 'New Delhi'}

**Summarization**

The summarization pipeline API has the ability to generate a summary of the given input text by keeping most of the important aspects

In [ ]:
from transformers import pipeline

summarizer = pipeline('summarization')

No model was supplied, defaulted to sshleifer/distilbart-cnn-12-6 and revision a4f8f3e (https://huggingface.co/sshleifer/distilbart-cnn-12-6).
Using a pipeline without specifying a model name and revision in production is not recommended.


config.json:   0%|          | 0.00/1.80k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/1.22G [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

We have to enter the input text along with a max_length or min_length argument for the summary generation of the text.

In [ ]:
summarizer("""
America has changed dramatically during recent years. Not only has the number of
    graduates in traditional engineering disciplines such as mechanical, civil,
    electrical, chemical, and aeronautical engineering declined, but in most of
    the premier American universities engineering curricula now concentrate on
    and encourage largely the study of engineering science. As a result, there
    are declining offerings in engineering subjects dealing with infrastructure,
    the environment, and related issues, and greater concentration on high
    technology subjects, largely supporting increasingly complex scientific
    developments. While the latter is important, it should not be at the expense
    of more traditional engineering.

    Rapidly developing economies such as China and India, as well as other
    industrial countries in Europe and Asia, continue to encourage and advance
    the teaching of engineering. Both China and India, respectively, graduate
    six and eight times as many traditional engineers as does the United States.
    Other industrial countries at minimum maintain their output, while America
    suffers an increasingly serious decline in the number of engineering graduates
    and a lack of well-educated engineers.
""", max_length=40)

Your min_length=56 must be inferior than your max_length=40.
/usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1202: UserWarning: Unfeasible length constraints: `min_length` (56) is larger than the maximum possible length (40). Generation will stop at the defined maximum length. You should decrease the minimum length and/or increase the maximum length.
  warnings.warn(


[{'summary_text': ' The number of U.S. engineering graduates in traditional engineering disciplines has declined in recent years . China and India graduate six and eight times as many traditional engineers as does the United States .'}]

We can only enter the input text and omit the max_length or min_length argument for the summary generation of the text

In [ ]:
summarizer("""
America has changed dramatically during recent years. Not only has the number of
    graduates in traditional engineering disciplines such as mechanical, civil,
    electrical, chemical, and aeronautical engineering declined, but in most of
    the premier American universities engineering curricula now concentrate on
    and encourage largely the study of engineering science. As a result, there
    are declining offerings in engineering subjects dealing with infrastructure,
    the environment, and related issues, and greater concentration on high
    technology subjects, largely supporting increasingly complex scientific
    developments. While the latter is important, it should not be at the expense
    of more traditional engineering.
    Rapidly developing economies such as China and India, as well as other
    industrial countries in Europe and Asia, continue to encourage and advance
    the teaching of engineering. Both China and India, respectively, graduate
    six and eight times as many traditional engineers as does the United States.
    Other industrial countries at minimum maintain their output, while America
    suffers an increasingly serious decline in the number of engineering graduates
    and a lack of well-educated engineers.
""")

[{'summary_text': ' The number of U.S. engineering graduates in traditional engineering disciplines has declined in recent years . China and India graduate six and eight times as many traditional engineers as does the United States . Rapidly developing economies such as India, Europe and Asia, continue to encourage and advance the teaching of engineering .'}]

**From the pipeline, use any model**.

Let us implement the text generation pipeline object by using the GPT2 model

In [ ]:
from transformers import pipeline

generator = pipeline('text-generation', model='gpt2')

config.json:   0%|          | 0.00/665 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/548M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/1.04M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

We can verify the model in the following way:

In [ ]:
generator.model.name_or_path

'gpt2'

**Tokenizers**

In [ ]:
!pip install transformers[sentencepiece] --q

In [ ]:
from transformers import AutoTokenizer

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

Here, the AutoTokenizer class has been imported from the transformers library and initialized with the model checkpoint name

In [ ]:
tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path=checkpoint)

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

Now, we have initialized the tokenizer.

After the execution of the code, we downloaded the tokenizer of the model named distilbert-base-uncased-finetuned-sst-2-english and cached it for further usage

In [ ]:
raw_inputs = [
    "I have been waiting for this HAAI course for last 3 months.",
    "I am very excited about learning new AI models !!",
    "I hate this hot weather which makes me feel irritated  !"
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[  101,  1045,  2031,  2042,  3403,  2005,  2023,  5292,  4886,  2607,
          2005,  2197,  1017,  2706,  1012,   102],
        [  101,  1045,  2572,  2200,  7568,  2055,  4083,  2047,  9932,  4275,
           999,   999,   102,     0,     0,     0],
        [  101,  1045,  5223,  2023,  2980,  4633,  2029,  3084,  2033,  2514,
         15560,   999,   102,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0]])}


Transformers only accept tensors, and that's why raw texts are first converted into tensors.

After checking the output, we can see input_ids consist of tokenized inputs and attention_mask contains the 1s and 0s which are used by the attention layer in the transformer defining whether the corresponding token should be attended or not.

**Model**

In [ ]:
from transformers import AutoModel

checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

The pre-trained model has been downloaded the same way as the tokenizer.

For loading the pre-trained model, the same from_pretrained method from the AutoModel class has been used

In [28]:
model = AutoModel.from_pretrained(checkpoint)

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

In [29]:
outputs = model(**inputs)
print(outputs.last_hidden_state.shape)

torch.Size([3, 16, 768])


We have the outputs with the help of tokenized inputs into the model. From the shape of the output’s hidden state, we observe that our model has a batch size (the number of sequences processed at a time) of 3, a sequence length (the length of each sequence) of 16, and a hidden state (the vector dimension of model input) of size 768

**Sequence Classification**

In [30]:
import torch
from transformers import AutoModelForSequenceClassification, AutoTokenizer
checkpoint = "distilbert-base-uncased-finetuned-sst-2-english"

In [31]:
tokenizer = AutoTokenizer.from_pretrained(pretrained_model_name_or_path=checkpoint)

Here we are downloading the model to get some predictions

In [32]:
model = AutoModelForSequenceClassification.from_pretrained(checkpoint)

In [33]:
raw_inputs = [
    "I have been waiting for this HAAI course for last 3 months.",
    "I am very excited about learning new AI models !!",
    "I hate this hot weather which makes me feel irritated  !"
]
inputs = tokenizer(raw_inputs, padding=True, truncation=True, return_tensors="pt")
print(inputs)

{'input_ids': tensor([[  101,  1045,  2031,  2042,  3403,  2005,  2023,  5292,  4886,  2607,
          2005,  2197,  1017,  2706,  1012,   102],
        [  101,  1045,  2572,  2200,  7568,  2055,  4083,  2047,  9932,  4275,
           999,   999,   102,     0,     0,     0],
        [  101,  1045,  5223,  2023,  2980,  4633,  2029,  3084,  2033,  2514,
         15560,   999,   102,     0,     0,     0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0]])}


Our tokenized inputs are ready. Let's go for the predictions

In [34]:
outputs = model(**inputs)
print(outputs)

SequenceClassifierOutput(loss=None, logits=tensor([[ 2.6973, -2.2845],
        [-4.0478,  4.3824],
        [ 3.9657, -3.2164]], grad_fn=<AddmmBackward0>), hidden_states=None, attentions=None)


In [35]:
print(outputs.logits.shape)

torch.Size([3, 2])


Our model has a matrix dimension of 3 x 2, as we have 3 sequences in the input and there are 2 classes.

We are now finding the outputs by passing them through the softmax activation to get the probabilities of each class for the input sentences

In [36]:
#Find the label / class probabilities
import torch
outputs = torch.nn.functional.softmax(outputs.logits, dim = -1)
print(outputs)

tensor([[9.9319e-01, 6.8147e-03],
        [2.1812e-04, 9.9978e-01],
        [9.9924e-01, 7.5948e-04]], grad_fn=<SoftmaxBackward0>)


We get [0.99, 0.006] as the output for the first input, [0.0002, 0.99] as the output for the second input, and finally [0.99, 0.0007] as the output for the third input sample.

Here we observe that our model is 99% confident that the first input sample belongs to the NEGATIVE class, 99% confident that the second input sample belongs to the POSITIVE class, and 99% confident that the third input sample belongs to the NEGATIVE class. We observe that the model’s output is quite accurate.

We can check the labels of the model in the following way:

In [37]:
model.config.id2label

{0: 'NEGATIVE', 1: 'POSITIVE'}